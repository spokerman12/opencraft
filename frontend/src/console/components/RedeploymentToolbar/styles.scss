@import '~styles/theme';


.redeployment-toolbar {
  background-color: $light-blue;
  padding: 15px;
  margin: 0 -15px;

  display: flex;
  align-items: center;

  .redeployment-nav {
    display: inline-block;
    width: 100%;
    max-width: 1100px;
  }
}

.tooltip {
  .arrow:before {
    border-right-color: white;
    color: white;
  }
  .tooltip-inner {
    padding: 10px;
    box-shadow: 0 4px 16px 0 rgba(0, 0, 0, 0.25);
    background-color: white;
    font-family: $chivo-font;
    font-size: 16px;
    letter-spacing: 0.44px;
    color: $secondary-1;
  }
  &.show {
    opacity: 1;
  }
}

.cancel-redeployment-modal {
  text-align: center;

  .modal-header {
    border-bottom: none;
    font-family: $playfair-font;
    font-size: 24px;
    color: #0f1f24;
    margin-bottom: -15px
  }

  .modal-body {
    background-color: #d3e1e7;
    margin: 20px;
    font-family: $chivo-font;
    font-size: 16px;
    color: #0f1f24;
  }

  .modal-footer {
    border: none;
    justify-content: center;
    margin-bottom: 15px;
  }
}


.div {
  z-index:2;
  position: absolute;
  right:10px;
  top: 0px;

  .toast {
    z-index:2;
    position:relative;
    top: 100px;
    height: auto;
    width: auto;
    box-shadow:  0 0 15px 0 rgba(0, 0, 0, 0.3);
    line-height: normal;
    background-color: white;
    border-left: 5px solid #ff9b04;

    .toast-header {
      display:block;
      height:0px;
      justify-content: flex-end;
      background-color: white;
      border-bottom: none;
    }

    .close {
      margin-top: 10px;
      justify-content: flex-end;
      font-size: 16px;
    }

  }
}
