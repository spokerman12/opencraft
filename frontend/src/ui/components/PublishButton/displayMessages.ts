const messages = {
  deploy: {
    defaultMessage: 'Publish',
    description: 'Deployment/Publish button text.'
  },
  notificationText: {
    defaultMessage: '{notificationNumber}',
    description: ''
  }
};

export default messages;
