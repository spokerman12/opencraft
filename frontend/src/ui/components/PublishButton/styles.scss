.notificationLayer {
    position: relative;
    margin-top: -25px;
    left: 70px;
    z-index:2;
    background-color:rgb(255, 0, 0) !important;
    height: 24px;
    width: 24px;
    border-radius: 100%;
    overflow:  visible;
    border-style: solid;
    border-color: white;
    border-width: 2px;
    box-shadow: 2px 2px #272727;
  }
  .notificationLayer p {
    display: inline;
    text-align: center;
    vertical-align: middle;
    font-weight: bold;
    font-size: 10px;
    letter-spacing: 0px;
    padding-top: 6px;
  }

  .textLayer {
    z-index:1;
    margin-top: 0px;
  }

  .disabledBtn {
    background-color:rgb(255, 255, 255) !important;
    border-color: grey !important;
    color:grey !important;
  }
