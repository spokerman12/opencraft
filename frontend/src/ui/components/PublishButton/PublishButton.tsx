import * as React from 'react';
import { WrappedMessage } from 'utils/intl';
import { Button } from 'react-bootstrap';
import { OpenEdXInstanceDeploymentStatusStatusEnum as DeploymentStatus } from 'ocim-client';
import { DeploymentInfoModel } from 'console/models';

import messages from './displayMessages';
import './styles.scss';

interface Props {
  deployment?: DeploymentInfoModel;
  loading: boolean;
  onClickWrapper: Function;
}

export const PublishButton: React.FC<Props> = ({
  deployment,
  loading,
  onClickWrapper
}: Props) => {
  let deploymentDisabled: boolean = true;
  let undeployedChanges: number = 0;
  let deploymentStatus: DeploymentStatus | null = null;
  if (deployment) {
    deploymentStatus = deployment.status;
    undeployedChanges = deployment.undeployedChanges.length;

    deploymentDisabled =
      loading ||
      !undeployedChanges ||
      deploymentStatus === DeploymentStatus.Preparing;
  }

  // Notifications red blip
  const notificationNumber = undeployedChanges <= 9 ? undeployedChanges : '9+';

  return (
    <Button
      className={
        deploymentDisabled
          ? 'float-right disabledBtn'
          : 'float-right enabledBtn'
      }
      variant="primary"
      size="lg"
      disabled={deploymentDisabled}
      onClick={() => {
        onClickWrapper();
      }}
    >
      {undeployedChanges !== 0 && (
        <div className="notificationLayer">
          <p>
            <WrappedMessage
              id="notificationText"
              messages={messages}
              values={{ notificationNumber }}
            />
          </p>
        </div>
      )}
      <div className="textLayer">
        <WrappedMessage id="deploy" messages={messages} />
      </div>
    </Button>
  );
};
