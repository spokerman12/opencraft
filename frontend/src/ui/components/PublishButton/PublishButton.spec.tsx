import React from 'react';
import { setupComponentForTesting } from "utils/testing";
import { RedeploymentToolbar } from 'console/components/RedeploymentToolbar';
import {
    OpenEdXInstanceDeploymentStatusDeploymentTypeEnum,
    OpenEdXInstanceDeploymentStatusStatusEnum
} from 'ocim-client';
import { PublishButton } from "./PublishButton"
/**
 * Returns the PublishButton's test props.
 *
 * @param {Object}  component - The root component.
 * @return {Array<String>} The specified props as an array of strings
 */
function getTestProps(publishBtn: Object): Array<String> {
  const disabled = publishBtn.props["disabled"]
  const notificationCount = publishBtn.children[0].props.className === "notificationLayer"
    ? publishBtn.children[0].children[0].children[0]
    : null
  return [disabled, notificationCount]
}

const singleDigitChanges = Array.from('x'.repeat(5)).length
const doubleDigitChanges = Array.from('x'.repeat(15)).length

it('renders without crashing', () => {
    const tree = setupComponentForTesting(<PublishButton />).toJSON();
    expect(tree).toMatchSnapshot();
});

describe("PublishButton renders as disabled when", function() {
  it('pending changes', () => {
      const publishBtn = setupComponentForTesting(
      <PublishButton
        undeployedChanges={singleDigitChanges}
        deploymentDisabled={true}
      />
      ).toJSON();
      expect(getTestProps(publishBtn)).toEqual([true, "5"])
  });

  it('no pending changes', () => {
    const publishBtn = setupComponentForTesting(
    <PublishButton
    undeployedChanges={[].length}
    deploymentDisabled={true}
    />
    ).toJSON();
    expect(getTestProps(publishBtn)).toEqual([true, null])
  });
});

describe("PublishButton renders as enabled when", function() {
  it('single-digit pending changes', () => {
    const publishBtn = setupComponentForTesting(
    <PublishButton
      undeployedChanges={singleDigitChanges}
      deploymentDisabled={false}
    />
    ).toJSON();
    expect(getTestProps(publishBtn)).toEqual([false, "5"])
  });
  it('double-digit pending changes', () => {
    const publishBtn = setupComponentForTesting(
    <PublishButton
      undeployedChanges={doubleDigitChanges}
      deploymentDisabled={false}
    />
    ).toJSON();
    expect(getTestProps(publishBtn)).toEqual([false, "9+"])
  });
})
