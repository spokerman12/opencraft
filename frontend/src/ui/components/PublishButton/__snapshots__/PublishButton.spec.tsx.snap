// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<button
  className="float-right enabledBtn btn btn-primary btn-lg"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  <div
    className="notificationLayer"
  >
    <p>
      9+
    </p>
  </div>
  <div
    className="buttonTextLayer"
  >
    Publish changes
  </div>
</button>
`;
